failure-injection: add disk stall failure modes

DarrylWong opened the pull request:
This change adds cgroup and dmsetup disk stalls to the failure injection library. A majority of this logic is a port of the existing disk stall implementations found in roachtestutil, however several additions were added to make cleanup and restore of said failures restore the system back to it's original state.

Informs: https://github.com/cockroachdb/cockroach/issues/138970
Release note: none

DarrylWong committed:
failure-injection: add disk stall failure modes

This change adds cgroup and dmsetup disk stalls to the failure
injection library. A majority of this logic is a port of the
existing disk stall implementations found in roachtestutil,
however several additions were added to make cleanup and
restore of said failures restore the system back to it's
original state.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143104)
<!-- Reviewable:end -->

DarrylWong commented on pkg/roachprod/failureinjection/failures/disk_stall.go:
```diff
@@ -0,0 +1,452 @@
+		return err
+	}
+
+	// N.B. Although the cgroupsv2 documentation states that "limits are in the range [0, max]",
```
Mystery solved after spending way too long reading reading the source code for the wrong distribution of Linux 🤣 Well, kinda 😅 . I assume 0 isn't allowed to avoid division by 0 when it's calculating the allowed bytes. No clue why 1 isn't allowed.

DarrylWong commented on pkg/roachprod/failureinjection/failures/disk_stall.go:
```diff
@@ -0,0 +1,466 @@
+		return err
+	}
+
+	// When we unmounted the disk in setup, the cgroups controllers may have been removed, re-add them.
```
This feels iffy to me to have to manually add back the controllers, but no matter how hard I tried I couldn't get them to be added back short of restarting the roachprod cluster.

srosenberg replied:
> I tried I couldn't get them to be added back short of restarting the roachprod cluster.

By "restarting", you mean VMs (not systemd)? It does look error-prone. What if we miss something; what's the chance that the cluster remains fully reusable (by independent roachtest)?

DarrylWong commented on pkg/roachprod/failureinjection/failures/disk_stall.go:
```diff
@@ -0,0 +1,466 @@
+	// If true, allow the failure mode to restart nodes as needed. E.g. dmsetup requires
+	// the cockroach process to not be running to properly setup. If RestartNodes is true,
+	// then the failure mode will restart the cluster for the user.
+	RestartNodes bool
```
The current disk stall roachtests are okay with the node dying since they don't try to do anything else in the test after the fact. However, I still think we should restart the node at the end so artifacts can be collected - you can see in all of the current roachtests there's no cockroach logs for the stalled node.

DarrylWong commented on pkg/roachprod/failureinjection/failures/disk_stall.go:
```diff
@@ -0,0 +1,466 @@
+	var err error
+
+	// Disabling journaling requires the cockroach process to not have been started yet.
+	if diskStallArgs.RestartNodes {
```
Note to self: having to restart the cluster to setup/cleanup dmsetup seems really disruptive - makes me lean towards using cgroups in the failure injection framework

srosenberg commented on pkg/cmd/roachtest/tests/failure_injection.go:
```diff
@@ -93,6 +102,11 @@ func (t *failureSmokeTest) run(
 	}
 	l.Printf("%s: Running Restore(); details in %s.log", t.failureName, file)
+	if err = c.AddGrafanaAnnotation(ctx, l, grafana.AddAnnotationRequest{
+		Text: fmt.Sprintf("%s restored", t.testName),
```
Nit: I probably mentioned this before... terminology-wise, `recovered` seems more unequivocal than `restored`, since the latter is usually used in the context of backup/restore.

DarrylWong replied:
Sure, renamed. I also have no strong preference which is why I just went the easy route of looking at what chaos mesh used :)

srosenberg commented on pkg/cmd/roachtest/tests/failure_injection.go:
```diff
@@ -331,33 +349,247 @@ var latencyTest = func(c cluster.Cluster) failureSmokeTest {
+		// Drop the page caches on each VM to force them to read from disk. Without
+		// this, we don't run the workload long enough for the process to read
+		// from disk instead of memory.
+		c.Run(ctx, option.WithNodes(c.CRDBNodes()), "echo 3 | sudo tee /proc/sys/vm/drop_caches")
```
This reminded of a really nice util (`vmtouch`) to managing and observing page cache [1].

[1] https://hoytech.com/vmtouch/

srosenberg commented on pkg/cmd/roachtest/tests/failure_injection.go:
```diff
@@ -331,33 +349,247 @@ var latencyTest = func(c cluster.Cluster) failureSmokeTest {
+		if err != nil {
+			return readWriteBytesOverTime{}, err
+		}
+		// Drop the page caches on each VM to force them to read from disk. Without
```
Nit: `3` also drops file cache, i.e., directory entry and inodes.

srosenberg commented on pkg/cmd/roachtest/tests/failure_injection.go:
```diff
@@ -331,33 +349,247 @@ var latencyTest = func(c cluster.Cluster) failureSmokeTest {
+	var tests []failureSmokeTest
+	for _, stallWrites := range []bool{true, false} {
+		for _, stallReads := range []bool{true, false} {
+			if !stallWrites && !stallReads {
```
I wonder if it might be still useful as a smoke test to verify that the no-op versions of `validateFailure` and `validateRestore` actually work as expected.

srosenberg commented on pkg/cmd/roachtest/tests/failure_injection.go:
```diff
@@ -331,33 +349,247 @@ var latencyTest = func(c cluster.Cluster) failureSmokeTest {
+			stalledNode := nodes[0]
+			unaffectedNode := nodes[1]
+
+			testName := fmt.Sprintf("%s: WritesStalled=%t ReadsStalled=%t", failures.CgroupsDiskStallName, stallWrites, stallReads)
```
Nit: maybe use `/` as the delimiter in case these later become top-level roachtests.

srosenberg commented on pkg/cmd/roachtest/tests/failure_injection.go:
```diff
@@ -331,33 +349,247 @@ var latencyTest = func(c cluster.Cluster) failureSmokeTest {
+					// (throughput is not a hard limit and limits can accumulate for bursts of io) we
+					// sometimes see some throughput.
+					if stallWrites {
+						if res.unaffectedNodeWriteBytes < res.stalledNodeWriteBytes*10 {
```
Is 10x over 30 seconds a safe threshold? I'm thinking it's workload-specific... lemme circle back to it, after having digested the whole PR.

srosenberg commented on pkg/cmd/roachtest/tests/failure_injection.go:
```diff
@@ -331,33 +349,247 @@ var latencyTest = func(c cluster.Cluster) failureSmokeTest {
+					// TODO(darryl): The failure mode itself should do this in WaitForFailureToRestore.
+					// It should also wait for replicas to rebalance, although this test is not large
+					// enough for that to matter.
+					db := c.Conn(ctx, l, stalledNode)
```
Don't we want `ConnE`? If node is unresponsive, `Conn` err is fatal.

DarrylWong committed:
failure-injection/roachtest: stall random amount of nodes instead of just one

This will exercise the failure injection library's ability to inject
failures on multiple nodes at once.

To support this, a SeededRandGroups helper was added to NodeListOption.

DarrylWong committed:
failure-injection: rename restore to recover

While the two names have similar meanings in the context of failure
injection, we often use restore to refer to restoring a backup. Lets
remove any ambiguity by renaming it to recover.

DarrylWong commented on pkg/roachprod/failureinjection/failures/failure.go:
```diff
@@ -107,3 +121,129 @@ func (f *GenericFailure) NetworkInterfaces(
+	return errors.CombineErrors(err, res[0].Err)
+}
+
+func (f *GenericFailure) WaitForSQLLiveness(
```
👍 Changed the name to `WaitForSqlReady`. I also added a `retryForDuration` helper that should extract the retry logic out and be reusable.

>I also wonder if we could merge it with WaitForSQLReady? 

Do you mean the `roachtestutil` `WaitForSQLReady`? Not aware of something similar on the roachprod layer.

srosenberg replied:
Yep. We can refactor on subsequent passes.

herkolategan reviewed (commented):
Nice work, left a few comments. Also curious if the cgroup test would work with the dmsetup staller as both seem to wait for and recover from node death?

DarrylWong commented:
>Also curious if the cgroup test would work with the dmsetup staller as both seem to wait for and recover from node death?

Unfortunately the cgroups test works by parsing `/sys/fs/cgroup/system.slice/io.stat` but `dmsetup` disables cgroups so that file doesn't exist.

Switching to `iotop` and filtering by PID should work for both, but I'll have to experiment with it. Fwiw though, I think what the `dmsetup` test does is a lot simpler/more intuitive, i.e. `if writes are stalled test that we can't write a file` seems like a stronger test than `heuristically checking disk is stalled because we observed "low enough" bytes`

DarrylWong commented on pkg/roachprod/failureinjection/failures/disk_stall.go:
```diff
@@ -0,0 +1,475 @@
+	})
+}
+
+func (s *CGroupDiskStaller) WaitForFailureToPropagate(
```
>I'm assuming it will always be set to a value will cause the node to die?

Good question, I don't think that's true so maybe always waiting for the node to die is too strong. The `registerDiskBandwidthOverload` for example, doesn't want the node to die it just slows writes to 128 MiB to test AC.

I have two trains of though for fixing this. The first is that we can just add a `ExpectNodeDeath` flag. The second is that we don't need to do anything, tests that don't expect the node to die can just skip calling `WaitForFailureToPropagate`. 

My thinking is that `WaitForFailureToPropagate` would be used mostly a suggestion anyway (i.e. log a warning but don't fail the test). Waiting for a disk stall to be detected (when the throughput isn't 0) or waiting for ranges to rebalance is workload dependent/not bounded anyway. e.g. if we slow writes to 5 MiB, it's not obvious if the node will die or not, so wait up to N minutes and log a warning if it doesn't.

DarrylWong commented on pkg/roachprod/failureinjection/failures/disk_stall.go:
```diff
@@ -0,0 +1,475 @@
+	return nil
+}
+
+func getStallType(diskStallArgs DiskStallArgs) ([]bandwidthType, error) {
```
Done.

DarrylWong commented on pkg/roachprod/failureinjection/failures/disk_stall.go:
```diff
@@ -0,0 +1,475 @@
+	// While this is exactly what  we want, it's not the intended use case and is invalid.
+	bytesPerSecond := 4
+	if diskStallArgs.Throughput > 0 {
+		bytesPerSecond = diskStallArgs.Throughput
```
Good call, I also changed the default to be 2. From what I can tell there's no reason to not set it to 2, I was just porting over the existing behavior.

srosenberg commented on pkg/roachprod/failureinjection/failures/failure.go:
```diff
@@ -107,3 +121,134 @@ func (f *GenericFailure) NetworkInterfaces(
+	return errors.Wrapf(err, "never connected to node %d after %s", node, timeout)
+}
+
+func (f *GenericFailure) WaitForSQLNodeDeath(
```
Nit: `WaitForSQLUnavailable`. "Death" implies the process is down, which `PingNode` can't establish; so, the current name is somewhat misleading.

srosenberg commented on pkg/roachprod/failureinjection/failures/disk_stall.go:
```diff
@@ -0,0 +1,474 @@
+	return &CGroupDiskStaller{GenericFailure: genericFailure}, nil
+}
+
+const CgroupsDiskStallName = "cgroup-disk-stall"
```
Nit: move above the `func` s.t. all var/const/type declarations precede the definitions.

srosenberg commented on pkg/roachprod/failureinjection/failures/disk_stall.go:
```diff
@@ -0,0 +1,474 @@
+	return CgroupsDiskStallName
+}
+
+type DiskStallArgs struct {
```
Nit: move above the `func` s.t. all var/const/type declarations precede the definitions.

srosenberg commented on pkg/roachprod/failureinjection/failures/disk_stall.go:
```diff
@@ -0,0 +1,474 @@
+		if err := s.Run(ctx, l, s.c.Nodes, "mkdir -p {store-dir}/logs"); err != nil {
+			return err
+		}
+		if err := s.Run(ctx, l, s.c.Nodes, "rm -f logs && ln -s {store-dir}/logs logs || true"); err != nil {
```
This would end up wiping node logs, every time this failure is injected, for the duration of some arbitrary test.

DarrylWong replied:
Ah actually I see, it's deleting the symlink pointer file. Although if I understand my order of operations the previous implementation was wrong since it would hit the above error described and silently fail.

srosenberg commented on pkg/roachprod/failureinjection/failures/disk_stall.go:
```diff
@@ -0,0 +1,474 @@
+	if diskStallArgs.StallWrites {
+		stallTypes = []bandwidthType{writeBandwidth}
+	}
+	if diskStallArgs.StallReads {
```
Should we further validate if `StallLogs` is set _without_ `StallWrites`? In that case, we don't expect the process to panic.

srosenberg commented on pkg/roachprod/failureinjection/failures/disk_stall.go:
```diff
@@ -0,0 +1,474 @@
+		`sudo dmsetup create data1`); err != nil {
+		return err
+	}
+	// This has occasionally been seen to fail with "Device or resource busy",
```
Nit: missing TODO?

DarrylWong replied:
This was copy and pasted from the existing disk stall code, i'm not too familiar with the context unfortunately.

srosenberg commented on pkg/roachprod/failureinjection/failures/disk_stall.go:
```diff
@@ -0,0 +1,474 @@
+	// snapd will run "snapd auto-import /dev/dm-0" via udev triggers when
+	// /dev/dm-0 is created. This possibly interferes with the dmsetup create
+	// reload, so uninstall snapd.
+	if err = s.Run(ctx, l, s.c.Nodes, `sudo apt-get purge -y snapd`); err != nil {
```
This makes me wonder if we should disable cluster reuse for anything that touches system dependencies. `apt` tends to flake occasionally and adds a source of non-determinism.

srosenberg commented on pkg/roachprod/failureinjection/failures/disk_stall.go:
```diff
@@ -0,0 +1,474 @@
+	return nil
+}
+
+func (s *DmsetupDiskStaller) WaitForFailureToPropagate(
```
Both of these helpers seem generic; could be moved to `GenericFailure`.

DarrylWong replied:
Right now they are generic but in the future I plan on adding checks for e.g. replication, range rebalancing, etc.

srosenberg commented on pkg/cmd/roachtest/tests/failure_injection.go:
```diff
@@ -324,58 +343,312 @@ var latencyTest = func(c cluster.Cluster) failureSmokeTest {
+	stalledNode := stalledNodeGroup.SeededRandNode(rng)
+	unaffectedNode := unaffectedNodeGroup.SeededRandNode(rng)
+
+	ableToCreateFile := func(ctx context.Context, l *logger.Logger, c cluster.Cluster, node option.NodeListOption) bool {
```
Nit: `ableToCreateFile` -> `touchFile`. Since we can't tell exactly why this func failed, e.g., ssh timeout, we should use a "weaker", i.e., less precise, name. Otherwise, the name could mislead the author/reader into thinking that the `false` result is caused by something else entirely, e.g., file permissions.

srosenberg commented on pkg/cmd/roachtest/tests/failure_injection.go:
```diff
@@ -324,58 +343,312 @@ var latencyTest = func(c cluster.Cluster) failureSmokeTest {
+			}
+			return nil
+		},
+		workload: func(ctx context.Context, c cluster.Cluster) error {
```
Same workload is duplicated above, albeit with different CLI options.

srosenberg commented on pkg/cmd/roachtest/tests/failure_injection.go:
```diff
@@ -324,58 +343,312 @@ var latencyTest = func(c cluster.Cluster) failureSmokeTest {
 			if err := test.run(ctx, t.L(), c, fr); err != nil {
 				t.Fatal(errors.Wrapf(err, "%s failed", test.testName))
 			}
+			cancel()
```
For clean shutdown, we want this inside `defer` since `Fatal` escapes.

srosenberg commented on pkg/roachprod/failureinjection/failures/disk_stall.go:
```diff
@@ -0,0 +1,474 @@
+	// Check the number of bytes read and written to disk.
+	res, err := s.RunWithDetails(
+		ctx, l, node,
+		fmt.Sprintf(`grep -E '%d:%d' /sys/fs/cgroup/system.slice/io.stat |`, maj, min),
```
It seems there are some lightweight metrics collectors, e.g., [1], which we could easily add as a scrape target. This could make it easier to debug; rather than sampling over discrete intervals ourselves, we'd have a more or less continuous view in Grafana.

[1] https://github.com/arianvp/cgroup-exporter/tree/main/collector

srosenberg reviewed (approved):
Nice work! There are a number of things we could improve in subsequent PR(s). It might be instructive to document some of them by way of GH issues. Otherwise, feel free to merge after a final sanity pass. 🚢

DarrylWong commented on pkg/roachprod/failureinjection/failures/failure.go:
```diff
@@ -59,18 +71,20 @@ type GenericFailure struct {
 	cmd := strings.Join(args, " ")
+	l.Printf("running cmd: %s", cmd)
 	// In general, most failures shouldn't be run locally out of caution.
 	if f.c.IsLocal() {
```
Once we get an actual failure injection planner, I see value in being able to dry run failure modes. Although maybe that should be dependent on a `--dry-run` flag and not local cluster.

yuzefovich commented:
Looks like the rebase has gone wrong.

DarrylWong commented:
Unfortunately to stabilize the test I had to make the read assertion weaker. The cluster just doesn't read a consistent amount of bytes, even with the same seed. It seems to have made the past 5 runs pass though, so I'll merge this given most of the problems seem to lie in the smoke testing itself and not the failure mode. 

TFTRs! 

bors r=herkolategan, srosenberg
